[evanpy] iP

evanpy opened the pull request:
#### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

evanpy committed:
Added echo functionality

evanpy committed:
Added task adding and listing functionality

evanpy committed:
Added Task class that can be marked/unmarked

evanpy committed:
Added Task types

evanpy committed:
Modified testing files for automatic testing for bot

evanpy committed:
Modified test files

evanpy committed:
Added exception handling for unknown input and missing task description

evanpy committed:
Modified test files

evanpy committed:
Added constructor documentation

evanpy committed:
Added support for deleting tasks

evanpy committed:
Edit Javadoc of mark and unmark methods

evanpy committed:
Refactor code and modify code to comply with coding standard

evanpy committed:
Add method to get description of task

evanpy committed:
Add DukeException.class

evanpy committed:
Add data saving and loading feature

evanpy committed:
Add formatting to task dates

evanpy committed:
Add formatting to task dates

evanpy committed:
Fix bug resulting from parsing date

evanpy committed:
Modify testing files

evanpy committed:
Refactor code into different classes

evanpy committed:
Place classes into package

evanpy committed:
modify test file to correct package file path

evanpy committed:
organize file structure of classes and modify code accordingly

evanpy committed:
move classes to different directory

evanpy committed:
Merge branch 'master' into add-gradle-support

evanpy committed:
change mainClassName to duke path

evanpy committed:
update gitignore

evanpy committed:
rename Todo file to ToDo

evanpy committed:
rename todo file to ToDo

evanpy committed:
Amend javadoc param typo

evanpy committed:
Add JUnit tests for tasks

evanpy committed:
Add javadoc to Duke class

evanpy committed:
Edit code for coding standard

evanpy committed:
Add query search on task list

evanpy committed:
Merge branch 'branch-A-CodingStandard'

evanpy committed:
Add query search to find specific tasks in task list

evanpy committed:
Add checkstyle plugin

evanpy committed:
Amend code to comply with coding standard

evanpy committed:
Add javafx dependencies and change main class

evanpy committed:
Modify code to replace text UI with GUI

evanpy committed:
Add fxml for user interface

evanpy committed:
Turn image frame to circular

rexcyrio commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+
+    /**
+     * You should have your own function to generate a response to user input.
+     * Replace this stub with your completed method.
```
Is this JavaDoc for this method? It sounds a little off.

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+
+import duke.task.ToDo;
+import duke.task.Deadline;
+import duke.task.Event;
```
Good job remembering to list each `import` explicitly!

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+		case "unmark":
+			output = tasks.unmarkTask(tasks.getTask(Integer.parseInt(descriptions[1]) - 1));
+			break;
+		case "todo":
```
Good job remembering to have the `case` statements on the same indentation level as the main `switch` statement!

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+	 * @param tasks List of tasks.
+	 * @param taskType Type of task.
+	 * @param descriptions The descriptions of the task.
+	 * @return Message containing task result.
```
Nicely formatted JavaDocs!
✅ Capital letters
✅ Empty line between description and parameter section
✅ Each line ends with a full stop

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+
+public class Parser {
+	/**
+	 * Creates task and stores it in Task arraylist.
```
Minor grammatical error:
> Creates **a** task and stores it in **the** Task **ArrayList**.

rexcyrio commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,43 @@
+    public String toString() {
+        return "[E]" + super.toString() 
+            + "(from: " + this.startTime.format(DateTimeFormatter.ofPattern("MMM d yyyy")) 
+            + " to: " + this.endTime.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
[Shouldn't line wraps be double the normal indentation (i.e., 8 spaces)?](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Indentation%20for%20wrapped%20lines%20should%20be%208%20spaces)

rexcyrio commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private boolean isInitializingData = true;
```
Good job naming the boolean attribute `is...` correctly!

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+			output = tasks.listTasks();
+			break;
+		case "mark":
+			output = tasks.markTask(tasks.getTask(Integer.parseInt(descriptions[1]) - 1));
```
What happens if the user types in `mark abc`? Can `Integer.parseInt` handle the string `"abc"`?

Also, I think the line `tasks.markTask(tasks.getTask(Integer.parseInt(descriptions[1]) - 1));` can be split up to improve readability:

```java
int taskNumber = Integer.parseInt(descriptions[1]) - 1;
Task task = tasks.getTask(taskNumber);
tasks.markTask(task);
```

rexcyrio commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,102 @@
+     * @return Success and add task detail message.
+     */
+    public static String addTaskMsg(Task task, int taskListSize) {
+        return "Got it. I've added this task:\n    "
+                + task + "\n"
+                + "Now you have " + taskListSize + (taskListSize == 1 ? " task " : " tasks ") + "in the list.\n";
```
This section is a little hard to read, especially since there's a ternary operator inside it. Maybe it can be split up into 2 parts?

rexcyrio reviewed (commented):
LGTM. Code quality is generally very good! No overly nested `if` statements or the like 👍 

Side note:
- I noticed that there was an empty `src/main/data/taskData.txt` file committed. Did you forget to remove it before pushing?
- There were some `.class` files committed. You can try Googling or asking on the forum for help on how to remove them from GitHub (I'm not very good with `git` myself).

rexcyrio commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+			output = tasks.unmarkTask(tasks.getTask(Integer.parseInt(descriptions[1]) - 1));
+			break;
+		case "todo":
+			try {
+				if (descriptions.length != 2) {
+					throw new DukeException("OOPS!!! The description of a todo cannot be empty.\n");
+				}
+
+				output = tasks.addTask(new ToDo(descriptions[1]));
+			} catch (DukeException error) {
+				output = Ui.errorMsg(error.getMessage());
+			}
```
Rather than have multiple `try-catch` blocks, maybe you can try

```java
try {
    switch (taskType) {
    case "list":
        // ...
        break;
    case "mark":
        // ...
        break;
    case "unmark":
        // ...
        break;

    /* remaining cases here */

    }
} catch (DukeException error) {
    output = Ui.errorMsg(error.getMessage());
}
```

to reduce code duplication.

rexcyrio reviewed (commented):
(continuation from previous review)

evanpy committed:
Update javadoc of getResponse

evanpy committed:
Change images for GUI

evanpy committed:
Add persona to response

niekis commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * A public constructor to initialize Duke instance.
+     *
+     * @param filePath Path of file.
```
I think you can be more specific here by stating the purpose of this file.

niekis commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.IOException;
+
+import duke.controller.MainWindow;
```
Good work avoiding wildcard imports! 
However, I feel that the duke.controller.MainWindow should not be grouped together with the javafx imports.

niekis commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
This line is longer than 80 char. Maybe you can phrase it as:
```
            FXMLLoader fxmlLoader = new FXMLLoader(
                    Main.class.getResource("/view/MainWindow.fxml"));
```
instead?

niekis commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+     */
+    protected void loadData(TaskList tasks) {
+        try {
+
```
Was this an accidental empty line?

niekis commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+            // Ui.errorMsg(error.getMessage());
+        } catch (FileNotFoundException error) {
+            tasks.finishInitialization();
+            // Ui.errorMsg(error.getMessage());
```
hmm you left in some commented out code here, was this an accident?

niekis commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+				output = Ui.errorMsg(e.getMessage());
+			}
+			break;
+		case "delete":
```
seems to me like this line is on the wrong indentation level

niekis commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+			output = tasks.removeTask(Integer.parseInt(descriptions[1]) - 1);
+			break;
+        case "find":
+			TaskList filteredTasks = new TaskList(tasks.filteredTaskList(descriptions[1]));
```
seems like this line is also on the wrong indentation level. I think you can try this instead?
```
        case "delete":
	    output = tasks.removeTask(Integer.parseInt(descriptions[1]) - 1);
	    break;
        case "find":
            TaskList filteredTasks = new TaskList(tasks.filteredTaskList(descriptions[1]));
```

niekis reviewed (commented):
Code LGTM and decently readible for the most part! Just some nits to be fixed. Additionally you might want to remove your .class files that might have been accidentally commited to git. You can check this [link](https://stackoverflow.com/questions/14251253/how-not-to-display-the-class-in-git) out to see how you can fix this issue using .gitignore!

evanpy committed:
Modify code for styling

evanpy committed:
Enable assertions for gradle file execution

evanpy committed:
TaskList class: add assertions

addTask and finishInitialization method should behave as
expected.

Assertion has been added to
- addTask to check that task list is not empty after a task has been added to list
- finishInitialization to check that it is not being called after duke
  has been initiated

evanpy committed:
Clean code for better code quality

evanpy committed:
Replace Parser class with duplicate

Parser file shows correct indentation within intellij but failed
checkstyle. Upon checking github, the indentations were all wrong.

The Parser java and class file were deleted and replaced with a
duplicate file.

evanpy committed:
Add javadoc for filteredTaskList

evanpy committed:
Merge pull request #1 from evanpy/branch-A-Assertions

Branch a assertions

evanpy committed:
Fix merge conflicts

evanpy committed:
Merge pull request #2 from evanpy/branch-A-CodeQuality

Branch a code quality

evanpy committed:
Merge pull request #3 from evanpy/branch-A-Streams

Add javadoc for filteredTaskList

SHni99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,152 @@
+				output = tasks.unmarkTask(tasks.getTask(Integer.parseInt(descriptions[1]) - 1));
+				break;
+			case "todo":
+				try {
```
Arrowhead pattern, probably is due to the wrong indentation implemented (4 spacing)

SHni99 reviewed (commented):
LGTM, overall clean code and good to go jyjy :)

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Add files via upload

evanpy committed:
Change background color

evanpy committed:
Merge branch 'master' of https://github.com/evanpy/ip

evanpy committed:
Update code to comply with checkstyle

evanpy committed:
Enable checkstyle

evanpy committed:
miscellaneous change

evanpy committed:
Miscellaneous change

evanpy committed:
Merge branch 'master' of https://github.com/evanpy/ip

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Update README.md

evanpy committed:
Fix file storage path

evanpy committed:
Modify bye command to end program

evanpy committed:
Add error handling for mark and unmark method when task index is unknown

evanpy committed:
Add new message when listing empty task list

evanpy committed:
Add error handling for parser

evanpy committed:
Merge branch 'master' of https://github.com/evanpy/ip

evanpy committed:
Change message when task list is empty

evanpy committed:
Add error handling for delete command

evanpy committed:
Fix index out of array bug in Parser method

evanpy committed:
Delete Ui.png

evanpy committed:
Add product image to docs

evanpy committed:
Add extension to sort task list by task name

evanpy committed:
Update README.md

evanpy committed:
Modify code to comply with checkstyle

evanpy committed:
Modify code to reallocate file storage access

evanpy committed:
Fix bug for image file name

evanpy committed:
Update README
